R/MSnExperiment-class.R

```diff
@@ -783,27 +1383,126 @@ setMethod("filterFile", "MSnExperiment", function(object, file) {
+        stop("'msLevel' must be numeric", call. = FALSE)
+    ps <- ProcessingStep("filterMz", list(mz = mz, msLevel. = msLevel., ...))
+    object@processingQueue <- c(object@processingQueue, list(ps))
+    object@processing <- c(object@processing,
+                           paste0("Filter: trim m/z [", mz[1], "..", mz[2],
+                                  "] on MS level(s) ",
+                                  paste0(unique(msLevel.), collapse = " "),
+                                  " [", date(), "]"))
+    object
+})
+
+#' @rdname MSnExperiment
+setMethod("filterPrecursorScan", "MSnExperiment",
+          function(object, acquisitionNum, ...) {
+              if (missing(acquisitionNum))
+                  return(object)
+              object <- object[.filterSpectraHierarchy(
+                  as.data.frame(object@spectraData), acquisitionNum),
+                  drop = FALSE]
+              object@processing <- c(
+                  object@processing,
+                  paste0("Filter: select parent/children scans for ",
+                         paste0(acquisitionNum, collapse = " "),
+                         " [", date(), "]"))
+              object
+          })
+
+#' @rdname MSnExperiment
+setMethod("filterPolarity", "MSnExperiment", function(object, polarity.) {
+    if (missing(polarity.))
+        return(object)
+    object <- object[polarity(object) %in% polarity., drop = FALSE]
+    object@processing <- c(object@processing,
+                           paste0("Filter: select spectra with polarity  ",
+                                  paste0(polarity., collapse = " "),
+                                  " [", date(), "]"))
+    object
+})
+
+#' @rdname MSnExperiment
+setMethod("filterRt", "MSnExperiment", function(object, rt, msLevel.) {
+    if (missing(rt)) return(object)
+    if (!(length(rt) == 2 & is.numeric(rt)))
+        stop("'rt' must be a numeric of length 2", call. = FALSE)
+    if (missing(msLevel.))
+        msLevel. <- unique(msLevel(object))
+    else if (!is.numeric(msLevel.))
+        stop("'msLevel' must be numeric", call. = FALSE)
+    rt <- range(rt)
+    msLevel. <- unique(msLevel.)
+    sel_ms <- msLevel(object) %in% msLevel.
+    sel_rt <- rtime(object) >= rt[1] & rtime(object) <= rt[2] & sel_ms
+    object <- object[sel_rt | !sel_ms, drop = FALSE]
+    object@processing <- c(object@processing,
+                           paste0("Filter: select retention time [",
+                                  rt[1], "..", rt[2], "] on MS level(s) ",
+                                  paste0(unique(msLevel.), collapse = " "),
+                                  " [", date(), "]"))
+    object
+})
+
+#' @rdname MSnExperiment
+setMethod("splitByFile", c("MSnExperiment", "factor"), function(object, f) {
+    if (length(f) != length(fileNames(object)))
+        stop("length of 'f' has to match the length of samples/files in 'object'.")
+    idxs <- lapply(levels(f), function(z) which(f == z))
```

Comment:
I would completely replace this method body by:
```r
    lapply(split(seq_along(f), f), filterFile, object=object)
```

See here some benchmarks:
```r
library("microbenchmark")

f <- as.factor(sample(letters[1:5], 10, replace=TRUE))
#  [1] c e e e a d e b d b
# Levels: a b c d e

f1 <- function(f)lapply(levels(f), function(z) which(f == z))
f2 <- function(f)split(seq_along(f), f)

all.equal(f1(f), f2(f))
# [1] "names for current but not for target"

microbenchmark(f1(f), f2(f))
# Unit: microseconds
#   expr    min      lq     mean  median      uq     max neval
#  f1(f) 37.535 38.2675 40.87978 38.8440 39.4545 107.486   100
#  f2(f)  4.932  5.5355  6.46035  6.0855  6.4645  30.464   100

# I used length instead of filterFile as example
f3 <- function(f) {
    idxs <- lapply(levels(f), function(z) which(f == z))
    res <- lapply(idxs, function(z) {
        length(z)
    })
    names(res) <- levels(f)
    res
}
f4 <- function(f) {
    lapply(split(seq_along(f), f), length)
}

all.equal(f3(f), f4(f))
# [1] TRUE

microbenchmark(f3(f), f4(f))
# Unit: microseconds
#   expr    min      lq     mean median      uq     max neval
#  f3(f) 36.231 37.6065 39.15445 38.002 38.5075 113.753   100
#  f4(f)  7.702  8.2400  8.99612  8.756  9.2070  32.700   100
```

Reply:
Agree 👍